src/main/scala/tech/phosphorus/intellij/prolog/formatting-defs.scala

```diff
@@ -0,0 +1,124 @@
+package tech.phosphorus.intellij.prolog.formatting
+
+import com.intellij.formatting._
+import com.intellij.lang.{ASTNode, Language}
+import com.intellij.psi.TokenType
+import com.intellij.psi.codeStyle.LanguageCodeStyleSettingsProvider.SettingsType
+import com.intellij.psi.codeStyle.{CodeStyleSettings, CodeStyleSettingsCustomizable, LanguageCodeStyleSettingsProvider}
+import com.intellij.psi.formatter.common.AbstractBlock
+import tech.phosphorus.intellij.prolog.PrologLanguage
+import tech.phosphorus.intellij.prolog.formatting.PrologFormattingModelBuilder.createSpaceBuilder
+import tech.phosphorus.intellij.prolog.psi.PrologTypes
+
+import java.util
+import scala.annotation.tailrec
+
+class PrologFormattingModelBuilder extends FormattingModelBuilder {
+  override def createModel(formattingContext: FormattingContext): FormattingModel = {
+    val codeStyleSettings = formattingContext.getCodeStyleSettings
+    FormattingModelProvider
+      .createFormattingModelForPsiFile(formattingContext.getContainingFile,
+        new RuleBlock(formattingContext.getNode,
+          Wrap.createWrap(WrapType.NONE, false),
+          Alignment.createAlignment(),
+          createSpaceBuilder(codeStyleSettings)),
+        codeStyleSettings)
+  }
+}
+
+object PrologFormattingModelBuilder {
+  def createSpaceBuilder(settings: CodeStyleSettings): SpacingBuilder = {
+    return new SpacingBuilder(settings, PrologLanguage.INSTANCE)
+      .between(PrologTypes.UNIFY, PrologTypes.DOT)
+      .spaceIf(settings.getCommonSettings(PrologLanguage.INSTANCE.getID).ALIGN_MULTILINE_METHOD_BRACKETS);
+  }
+}
+
+class RuleBlock(node: ASTNode, wrap: Wrap, alignment: Alignment, spacingBuilder: SpacingBuilder) extends AbstractBlock(node, wrap, alignment) {
+  override def buildChildren(): util.List[Block] = {
+    val blocks = new util.ArrayList[Block]()
+    addBlocks(node.getFirstChildNode, blocks)
+    return blocks
+  }
+
+  @tailrec
+  private def addBlocks(child: ASTNode, blocks: util.ArrayList[Block]): Unit = {
+    if (child != null) {
+      child.getElementType match {
+        case PrologTypes.UNIFY => wrapblock(child, blocks)
+        case PrologTypes.COMMA => wrapblock(child, blocks)
+        case PrologTypes.DOT => wrapblock(child, blocks)
+        case TokenType.WHITE_SPACE => wrapblock(child, blocks)
+        case non_white_space_rest => blocks.add(new RuleBlock(child, Wrap.createWrap(WrapType.NONE, false), Alignment.createAlignment(),
```

Comment:
I'm unsure if the `non_white_space_rest` is for annotating (which is not bad? I might say) or somehow. But you can always use a wildcard `_` instead of this unused wildcard value.